### ISSUE TITLE
rofi 0.15.12 is unusable with a custom keyboard layout

### ISSUE THREAD

**Issue by mardukbp** [DaveDavenport/rofi#307]
In my keyboard layout I mapped Alt to Caps Lock and Shift to Alt. So I do have an Alt key, that I use for emacs. However, the new version of rofi complains that my keyboard layout lacks an Alt key and I cannot use rofi anymore.

I read the release notes, but honestly I do not understand the rationale behind this behavior. It seems to me that rofi is being quite intrusive. Why does it care about the keyboard layout? Given that it is possible to configure rofi's keybindings, I do not see the need to check for a "sane" layout.

**Comment by DaveDavenport** [comment]
I need to know what keys in the state mask are modifier keys and what bit correspondence to what key.
F.e. I need to grab modfier keys but not ISO_* keys.
Say you have the right alt set to ISO_Level3_Shift for entering éáðßf,  there is no longer a AltGR key in the keymap,   the bit in the state field previously set to the AltGR key I now need to ignore when testing bindings.

This is why I made the key lookup dynamic.  I am sorry this now fails for you.  I *never* use different keyboard layouts (only us) so this is all very new for me, if you know how to handle this better please provide a patch. I am just trying to figure this out as I go, and I have not found a single complete/correct/up2date documentation that describes how to handle this.

Can you also include some steps on how to reproduce your issue?

**Comment by mardukbp** [comment]
Thank you very much for your detailed reply. After solving this problem I have a better understanding of how to define a custom keymap. I will summarize my findings below.

#### The problem

In the following I will refer to the function `x11_figure_out_numlock_mask` defined in `rofi/source/x11-helper.c`.

The problem was that `kc_altl` points to the correct keycode for the `Alt_L` key in my custom keymap, but this keycode is not found anywhere in `modmap->modifiermap`, which results in `AltMask` not being set to 1.

For debugging purposes, it is good to know that the keycodes stored in  `modmap->modifiermap` correspond to the output of the command `xmodmap`. In order to convert the hex codes to decimal
keycodes use e.g. `echo $((0x32))`.

#### xmodmap

In the following I will assume that the current keymap is vanilla `us`: `setxkbmap us -option`.

The current keymap can be exported to a file with `xmodmap -pke > xmodmaprc`. This file contains a dictionary mapping keycodes to keys. If one wants to e.g. use `Alt_L` as `Shift_L`, it is only a matter of changing the line corresponding to keycode 64 and then calling `xmodmap xmodmaprc`.

Now issuing `xmodmap` shows that no keycode is bound to `Alt_L` and `rofi` will obviously complain. How does one make `rofi` happy? Add the following lines at the BOTTOM of `xmodmaprc`:

```
clear mod1
add mod1 = Alt_L Alt_R Meta_L
```

Adding lines beginning with `clear` and `add` at the top of the file when a lot of keys have been remapped leads to anger->hate->suffering (that's what happened to me!).

Now running `xmodmap xmodmaprc && xmodmap` will show that there is an `Alt_L` bound to keycode 204 (0xcc), that is `Alt_R + Alt_L = Alt_L`, and `rofi` will not complain anymore.

#### Bottom line

Whenever one rebinds the keycodes of modifier keys in `xmodmaprc` one also has to update the modifier map. This is done using the commands `clear` and `add` as explained above.

**Comment by DaveDavenport** [comment]
Thanks for this writeup.  Is it ok if I turn this in to a wiki page later?

**Comment by mardukbp** [comment]
Sure. Thanks!

**Comment by carnager** [comment]
https://github.com/DaveDavenport/rofi/wiki/Custom-Keymaps
